### PR TITLE
executor: tiny refine for join

### PR DESCRIPTION
**Pull request by zz-jason**
fix https://github.com/pingcap/tidb/issues/4778
1. improve CPU efficiency:
    1. encode/decode inner row index instead of the whole inner row
    2. for inner join implemented in hash join, don't start join worker and outer table fetcher
2. improve MEM efficiency:
    1. for hash join, the hash map uses the encoded inner row id as value instead of the encoded inner row to save memory usage.
    2. for hash join, copy inner rows to a continuous memory to improve cache performance.
    3. for join result generators, pre-allocate a continuous memory to store all the joined row and employ `makeJoinRowToBuffer` to do the join work

### CODE REVIEW

**Comment by zz-jason** [comment]
@hanfei1991 @coocood @XuHuaiyu PTAL

**Comment by shenli** [comment]
@zz-jason Please resolve the conflicts.

**Comment by coocood** [comment]
@zz-jason 
If we don't encode the full row of inner, the used memory going to be much larger.

File: executor/join.go [hunk 1]
```diff
@@ -296,7 +299,47 @@ func (e *HashJoinExec) runJoinWorker(workerID int) {
...
+					j--
+				}
+				if i < j && outerFilterResult[i] == false && outerFilterResult[j] == true {
+					outerFilterResult[i], outerFilterResult[j] = outerFilterResult[j], outerFilterResult[i]
+					outerBuffer.rows[i], outerBuffer.rows[j] = outerBuffer.rows[j], outerBuffer.rows[i]
+					i++
+					j--
+				}
+			}
+			for numMatchedOuters < numOuters && outerFilterResult[numMatchedOuters] == true {
```
Comment (XuHuaiyu) [inline executor/join.go]: can we use `numMatchedOuters = i`

File: executor/join.go [hunk 2]
```diff
@@ -296,7 +299,47 @@ func (e *HashJoinExec) runJoinWorker(workerID int) {
...
+				numMatchedOuters++
+			}
+		} else {
+			numMatchedOuters = numOuters
+		}
+
+		if len(outerBuffer.rows)-numMatchedOuters > 0 {
```
Comment (XuHuaiyu) [inline executor/join.go]: I think we do not need this check,
if len(outerBuffer.rows)-numMatchedOuters == 0, `outerBuffer.rows[numMatchedOuters:]` will be `[]`

File: executor/join.go [hunk 3]
```diff
@@ -296,7 +299,47 @@ func (e *HashJoinExec) runJoinWorker(workerID int) {
...
+			resultBuffer.rows = e.resultGenerator.emitUnMatchedOuters(outerBuffer.rows[numMatchedOuters:], resultBuffer.rows)
+			if len(resultBuffer.rows) >= bufferCapacity {
```
Comment (XuHuaiyu) [inline executor/join.go]: This check can be removed, and move line 347 to line 343

File: executor/join.go [hunk 4]
```diff
@@ -296,7 +299,47 @@ func (e *HashJoinExec) runJoinWorker(workerID int) {
...
+					outerBuffer.rows[i], outerBuffer.rows[j] = outerBuffer.rows[j], outerBuffer.rows[i]
+					i++
+					j--
+				}
+			}
+			for numMatchedOuters < numOuters && outerFilterResult[numMatchedOuters] == true {
+				numMatchedOuters++
+			}
+		} else {
+			numMatchedOuters = numOuters
```
Comment (XuHuaiyu) [inline executor/join.go]: This can be moved to line 302
and set numMatchedOuters = 0 on line 304.

File: executor/join.go [hunk 5]
```diff
@@ -296,7 +299,47 @@ func (e *HashJoinExec) runJoinWorker(workerID int) {
...
+			}
+
+			for i, j := 0, len(outerFilterResult)-1; i < j; {
+				for i < j && outerFilterResult[i] == true {
+					i++
+				}
+				for i < j && outerFilterResult[j] == false {
+					j--
+				}
+				if i < j && outerFilterResult[i] == false && outerFilterResult[j] == true {
```
Comment (XuHuaiyu) [inline executor/join.go]: outerFilterResult[i] == false && outerFilterResult[j] == true can be promised.

File: executor/join.go [hunk 6]
```diff
@@ -296,7 +299,47 @@ func (e *HashJoinExec) runJoinWorker(workerID int) {
...
+					outerFilterResult[i], outerFilterResult[j] = outerFilterResult[j], outerFilterResult[i]
+					outerBuffer.rows[i], outerBuffer.rows[j] = outerBuffer.rows[j], outerBuffer.rows[i]
+					i++
```
Comment (XuHuaiyu) [inline executor/join.go]: if we use i<=j, and remove line 324 and line 325, i can be promised to end at `false`.

File: executor/join_result_generators.go [hunk 7]
```diff
@@ -319,32 +363,24 @@ type innerJoinResultGenerator struct {
...
-	originLen := len(resultBuffer)
-	for _, joinedRow := range joinedRows {
-		matched, err := expression.EvalBool(outputer.filter, joinedRow, outputer.ctx)
-		if err != nil {
-			return resultBuffer, false, errors.Trace(err)
-		}
-		if matched {
-			resultBuffer = append(resultBuffer, joinedRow)
+			buffer = outputer.makeJoinRowToBuffer(buffer[len(buffer):], outer, inner)
+			resultBuffer = append(resultBuffer, buffer)
```
Comment (XuHuaiyu) [inline executor/join_result_generators.go]: move this line to line382

File: executor/join_result_generators.go [hunk 8]
```diff
@@ -70,25 +70,64 @@ type baseJoinResultGenerator struct {
...
+	buffer = append(buffer, rhs...)
+	return buffer
+}
+
+// growResultBufferIfNecessary grows resultBuffer if necessary.
+func (outputer *baseJoinResultGenerator) growResultBufferIfNecessary(resultBuffer []Row, numToAppend int) []Row {
+	if cap(resultBuffer)-len(resultBuffer) >= numToAppend {
+		return resultBuffer
+	}
+	newResultBuffer := make([]Row, len(resultBuffer), len(resultBuffer)+numToAppend)
```
Comment (XuHuaiyu) [inline executor/join_result_generators.go]: extract len(resultBuffer) as a variable.

File: executor/join_result_generators.go [hunk 9]
```diff
@@ -157,21 +199,25 @@ type leftOuterSemiJoinResultGenerator struct {
 
 // emitMatchedInners implements joinResultGenerator interface.
 func (outputer *leftOuterSemiJoinResultGenerator) emitMatchedInners(outer Row, inners []Row, resultBuffer []Row) ([]Row, bool, error) {
+	if len(inners) == 0 {
```
Comment (XuHuaiyu) [inline executor/join_result_generators.go]: extract len(inners) and len(outers) as variables will be more readable
so as the others.

File: executor/join_result_generators.go (hunk 7 above)
Comment (XuHuaiyu) [reply executor/join_result_generators.go]: never mind

**Comment by XuHuaiyu** [comment]
LGTM

File: executor/join.go [hunk 10]
```diff
@@ -275,10 +280,43 @@ func (e *HashJoinExec) waitJoinWorkersAndCloseResultChan() {
...
 }
 
+// filterOuters filters the outer rows stored in "outerBuffer" and move all the matched rows ahead of the unmatched rows.
+// The number of matched outer rows is returned as the first value.
+func (e *HashJoinExec) filterOuters(outerBuffer *execResult, outerFilterResult []bool) (int, error) {
+	if e.outerFilter == nil {
+		return len(outerBuffer.rows), nil
+	}
+
+	outerFilterResult = outerFilterResult[:0]
```
Comment (tiancaiamao) [inline executor/join.go]: Could you use a one-pass algorithm and eliminate this auxiliary array?

**Comment by winoros** [comment]
Does 1.i still take effect?

File: executor/join.go [hunk 11]
```diff
@@ -228,45 +243,35 @@ func (e *HashJoinExec) prepare() error {
...
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	err = decodeRawValues(values, e.innerlExec.Schema(), e.ctx.GetSessionVars().GetTimeZone())
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	return values, nil
+	// start a goroutine to wait join workers finish their job and close channels.
+	go e.waitJoinWorkersAndCloseResultChan()
```
[the remark this answers is not in the document]
Comment (zz-jason) [reply executor/join.go]: It's not necessary, but it's the simplest and easiest way.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**